### PR TITLE
MX-13760: improve handlers sorting on epoch change subscriber

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
## Reasoning behind the pull request
- previously, the epoch start subscription handler sorted the handlers each time when notifying.
  
## Proposed changes
- when adding a new handler, sort the slice right after, so the handlers are always sorted

## Testing procedure
- regular testing procedure + check in logs that the NotifyOrder is followed

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: epochStart/notifier/epochStartSubscriptionHandler.go [hunk 1]
```diff
@@ -42,6 +42,9 @@ func (essh *epochStartSubscriptionHandler) RegisterHandler(handler epochStart.Ac
 	if handler != nil {
```
Comment (iulianpascalau) [inline epochStart/notifier/epochStartSubscriptionHandler.go:42]: please refactor here:
```GO
func (essh *epochStartSubscriptionHandler) RegisterHandler(handler epochStart.ActionHandler) {
    if check.IfNil(handler) {
        return
    }
....
```

File: epochStart/notifier/epochStartSubscriptionHandler.go [hunk 2]
```diff
@@ -62,11 +65,6 @@ func (essh *epochStartSubscriptionHandler) UnregisterHandler(handlerToUnregister
 // NotifyAll will call all the subscribed functions from the internal slice
 func (essh *epochStartSubscriptionHandler) NotifyAll(hdr data.HeaderHandler) {
 	essh.mutEpochStartHandler.RLock()
-
-	sort.Slice(essh.epochStartHandlers, func(i, j int) bool {
```
Comment (iulianpascalau) [inline epochStart/notifier/epochStartSubscriptionHandler.go:66]: :+1:

File: epochStart/notifier/epochStartSubscriptionHandler.go (hunk 1 above)
Comment (bogdan-rosianu) [reply epochStart/notifier/epochStartSubscriptionHandler.go]: done by check if nil reflect

File: state/syncer/userAccountsSyncer.go [hunk 3]
```diff
@@ -274,8 +274,8 @@ func (u *userAccountsSyncer) syncAccountDataTries(
 }
 
 func (u *userAccountsSyncer) printDataTrieStatistics() {
-	u.mutStatistics.RLock()
-	defer u.mutStatistics.RUnlock()
+	u.mutStatistics.Lock()
```
Comment (SebastianMarian) [inline state/syncer/userAccountsSyncer.go:277]: Why this change?

Comment (bogdan-rosianu) [reply state/syncer/userAccountsSyncer.go:277]: because there is a `sort.Slice(u.largeTries, func(i, j int) bool ` that happened under RLock/RUnlock. It's not really an issue since the `printDataTrieStatistics` function is called sequentially, but it's better to have the fix

Comment (SebastianMarian) [reply state/syncer/userAccountsSyncer.go:277]: ok

**Review by SebastianMarian — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed 
@@                            Log scanner                                  @@
##                            MX-13760-epoch-subscriber-improvements       ##
================================================================================
- Known Warnings              6
- New Warnings                1
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================
- block hash does not match                  9767
- wrong nonce in block                       3278
- miniblocks does not match                  0
- num miniblocks does not match              0
- miniblock hash does not match              0
- block bodies does not match                0
- receipts hash missmatch                    0
================================================================================
-  No jailed nodes on the testnet
================================================================================
